src/backend/dx11/src/lib.rs

```diff
@@ -1517,9 +1514,128 @@ impl CommandBufferState {
     }
 }
 
+fn generate_graphics_dynamic_constant_buffer_offsets<'a>(
+    bindings: impl IntoIterator<Item = &'a pso::DescriptorSetLayoutBinding>,
+    offset_iter: &mut impl Iterator<Item = u32>,
+    context1_some: bool,
+) -> (ArrayVec<[UINT; d3d11::D3D11_COMMONSHADER_CONSTANT_BUFFER_API_SLOT_COUNT as _]>, ArrayVec<[UINT; d3d11::D3D11_COMMONSHADER_CONSTANT_BUFFER_API_SLOT_COUNT as _]>) {
+    let mut vs_offsets = ArrayVec::new();
+    let mut fs_offsets = ArrayVec::new();
+
+    let mut exists_dynamic_constant_buffer = false;
+
+    for binding in bindings.into_iter() {
+        match binding.ty {
+            pso::DescriptorType::Buffer {
+                format: pso::BufferDescriptorFormat::Structured {
+                    dynamic_offset: true
+                },
+                ty: pso::BufferDescriptorType::Uniform,
+            } => {
+                let offset = offset_iter.next().unwrap();
+
+                if binding.stage_flags.contains(pso::ShaderStageFlags::VERTEX) {
+                    vs_offsets.push(offset / 16)
+                };
+
+                if binding.stage_flags.contains(pso::ShaderStageFlags::FRAGMENT) {
+                    fs_offsets.push(offset / 16)
+                };
+                exists_dynamic_constant_buffer = true;
+            }
+            pso::DescriptorType::Buffer {
+                format: pso::BufferDescriptorFormat::Structured {
+                    dynamic_offset: false
+                },
+                ty: pso::BufferDescriptorType::Uniform,
+            } => {
+                if binding.stage_flags.contains(pso::ShaderStageFlags::VERTEX) {
+                    vs_offsets.push(0)
+                };
+
+                if binding.stage_flags.contains(pso::ShaderStageFlags::FRAGMENT) {
+                    fs_offsets.push(0)
+                };
+            }
+            pso::DescriptorType::Buffer {
+                ty: pso::BufferDescriptorType::Storage { .. },
+                ..
+            } => {
+                // TODO: Storage buffer offsets require new buffer views with correct sizes.
+                //       Might also require D3D11_BUFFEREX_SRV to act like RBA is happening.
+                let _ = offset_iter.next().unwrap();
+                warn!("Dynamic offsets into storage buffers are currently unsupported on DX11.");
+            }
+            _ => {}
+        }
+    }
+
+    if exists_dynamic_constant_buffer && context1_some {
+        warn!("D3D11.1 runtime required for dynamic offsets into constant buffers. Offsets will be ignored.");
+    }
+
+    (vs_offsets, fs_offsets)
+}
+
+fn generate_compute_dynamic_constant_buffer_offsets<'a>(
+    bindings: impl IntoIterator<Item = &'a pso::DescriptorSetLayoutBinding>,
+    offset_iter: &mut impl Iterator<Item = u32>,
+    context1_some: bool,
+) -> ArrayVec<[UINT; d3d11::D3D11_COMMONSHADER_CONSTANT_BUFFER_API_SLOT_COUNT as _]> {
```

Comment:
can we has a typedef for this array vec? :)

Reply:
Done, the name of which is infinitely bikeshed-able 😄